src/main/java/com/commercetools/pspadapter/facade/CacheableSphereClientFactoryImpl.java

```diff
@@ -0,0 +1,53 @@
+package com.commercetools.pspadapter.facade;
+
+import com.commercetools.pspadapter.tenant.TenantConfig;
+import com.commercetools.pspadapter.util.CtpClientConfigurationUtils;
+import io.sphere.sdk.client.SphereClient;
+import io.sphere.sdk.client.SphereClientConfig;
+import org.springframework.cache.annotation.CacheConfig;
+import org.springframework.cache.annotation.Cacheable;
+import org.springframework.cache.annotation.EnableCaching;
+import org.springframework.stereotype.Component;
+
+import javax.annotation.Nonnull;
+
+/**
+ * {@link SphereClient} factory to allow caching sphere clients for {@link TenantConfig}s and {@link SphereClientConfig}.
+ * <p>
+ * This factory is a preferred way to create/get sphere clients for every tenant/sphereClient config
+ * <p>
+ * The main problem with non-cached clients is that they should be closed explicitly after an HTTP request, otherwise
+ * they become lost non-garbaged values. But closing the clients every time is quite complicated solution for current
+ * service implementation. Also, recreating sphere clients every time doesn't make much sense, so caching client for
+ * constant configs is a good approach.
+ * <p>
+ * <b>Note:</b> current solution doesn't support any cache eviction. This is not a problem so far since
+ * {@link SphereClient}s usually quite lightweight and we don't expect too much tenants on the same service. But if in
+ * the future the service grows to some monstrous application - eviction should be used. See more here:<ul>
+ * <li><a href="https://stackoverflow.com/questions/8181768/can-i-set-a-ttl-for-cacheable">Can I set a TTL for @Cacheable</a></li>
+ * <li><a href="https://github.com/ben-manes/caffeine/wiki/Eviction">Caffein Eviction</a></li>
+ * <li><b>Spring:</b><ul>
+ * <li><a href="https://docs.spring.io/spring/docs/3.1.x/spring-framework-reference/htmlsingle/spring-framework-reference.html#cache-specific-config">
+ * How can I set the TTL/TTI/Eviction policy/XXX feature?</a></li>
+ * <li><a href="https://docs.spring.io/spring/docs/current/spring-framework-reference/integration.html#cache">Cache Abstraction</a></li>
+ * <li><a href="https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/cache/annotation/CacheEvict.html">CacheEvict</a></li>
+ * </ul></li>
+ * </ul>
+ */
+@Component
+@EnableCaching
+@CacheConfig(cacheNames = "SphereClientFactoryCache")
+public class CacheableSphereClientFactoryImpl implements SphereClientFactory {
+
+    @Cacheable(sync = true)
+    @Override
+    public SphereClient createSphereClient(@Nonnull TenantConfig tenantConfig) {
```

Comment:
In `TenantConfig`, there's also `APIContextFactory` which is used to compare if object exists in cache or not. But don't you think it's redundant? Isn't it better to implement your own `equals` and `hashCode` than relying on the reflection?

Reply:
sorry, not clear what do you mean, why reflection equals/hash could is redundant here?

Reply:
`TenantConfig` contains `SphereClientConfig` and `APIContextFactory`. If you use @Cacheable, then it compares using `equals` and `equals` in TenantConfig compares also using `SphereClientConfig` and `APIContextFactory`. I find it a bit redundant.

Reply:
you mean the case when multiple tenant configs have the same CTP project credentials - they should fetch the same CTP client instance?

Reply:
but actually, this will work exactly like this, because [`CacheableSphereClientFactoryImpl#createSphereClient(TenantConfig)`](
https://github.com/commercetools/commercetools-paypal-plus-integration/blob/ad98559d28cc3a7c96c8bb9b20fd2bd75f793f94/src/main/java/com/commercetools/pspadapter/facade/CacheableSphereClientFactoryImpl.java#L44) calls [`CacheableSphereClientFactoryImpl#createSphereClient(SphereClientConfig)`](https://github.com/commercetools/commercetools-paypal-plus-integration/blob/ad98559d28cc3a7c96c8bb9b20fd2bd75f793f94/src/main/java/com/commercetools/pspadapter/facade/CacheableSphereClientFactoryImpl.java#L50), and the last one is also cacheable.

Reply:
Yes, it's correct. What I wanted to say is:
```java
TenantConfig a = new TenantConfig("tenantName","ctpProjectKey","ctpClientId","ctpClientSecret","pPlusClientId","pPlusClientSecret","pPlusClientMode")
TenantConfig b = new TenantConfig("tenantName","ctpProjectKey","ctpClientId","ctpClientSecret","pPlusClientId","pPlusClientSecret","pPlusClientMode")

a.equals(b) === false
```

Reply:
No, see the test: https://github.com/commercetools/commercetools-paypal-plus-integration/blob/a06763ac9f4172a235b15bcd0aa2c57b1ffd5bf4/src/test/java/com/commercetools/pspadapter/tenant/TenantConfigTest.java#L41

they are equal

Reply:
BUT, you were (partially, and guess unexpectedly for you) right!
https://stackoverflow.com/questions/16899604/spring-cache-cacheable-not-working-while-calling-from-another-method-of-the-s

calling cacheable from the same class doesn't work! and my new test fails:
```java
@Test
public void createSphereClient_whenDifferentTenantsHaveSameCtpConfig_returnsSameInstance() throws Exception {
    assertThat(cacheableSphereClientFactory).isNotNull();

    // different instances of the same config
    TenantConfig config1 = new TenantConfig("tenant_1", "ctpProjectKey1", "ctpClientId1", "ctpClientSecret1",
            "XXXXXXX", "XXXXXXX", SANDBOX);
    TenantConfig configSameCtpDifferentPp = new TenantConfig("tenant_2", "ctpProjectKey1", "ctpClientId1", "ctpClientSecret1",
            "YYYYYYY", "YYYYYYY", SANDBOX);

    assertThat(config1).isNotEqualTo(configSameCtpDifferentPp);
    assertThat(config1.getSphereClientConfig()).isEqualTo(config1.getSphereClientConfig());

    SphereClient client1 = cacheableSphereClientFactory.createSphereClient(config1.getSphereClientConfig());
    SphereClient client2 = cacheableSphereClientFactory.createSphereClient(configSameCtpDifferentPp.getSphereClientConfig());
    assertThat(client1).isNotNull();
    assertThat(client1).isSameAs(client2);

    client1 = cacheableSphereClientFactory.createSphereClient(config1);
    client2 = cacheableSphereClientFactory.createSphereClient(configSameCtpDifferentPp);
    assertThat(client1).isNotNull();
    assertThat(client1).isSameAs(client2);
}
```

Reply:
Aha, at least my effort was for sth :D

Reply:
ok, looks better now :), thanks!
d6a4d93eb3bad80cb056ee9fc21631594118c43f